File: golang/ingest/ingest_test.go
//go:build citenexus_ffi

package ingest

import (
	"encoding/json"
	"errors"
	"strings"
	"testing"

	"github.com/muthuishere/citenexus/golang/core"
)

// fakeEmbedder is a deterministic, hermetic stand-in for a real embedding model
// (CLAUDE.md: unit tests use deterministic fakes). It returns a fixed-dim vector
// whose first element is the chunk's word count, so different chunks differ.
type fakeEmbedder struct{ dim int }

func (f fakeEmbedder) Embed(text string) ([]float64, error) {
	vec := make([]float64, f.dim)
	vec[0] = float64(len(strings.Fields(text)))
	for i := 1; i < f.dim; i++ {
		vec[i] = float64((len(text) + i) % 7)
	}
	return vec, nil
}

// timeoutEmbedder is the real-world failure the seam exists for: the model dies
// on the Nth chunk. Before ADR-0014 R2 it had nowhere to say so and returned
// make([]float64, dim) — the zero vector ingest then indexed.
type timeoutEmbedder struct {
	dim    int
	calls  int
	failOn int
}

var errEmbedTimeout = errors.New("model call timed out after 30s")

func (e *timeoutEmbedder) Embed(text string) ([]float64, error) {
	e.calls++
	if e.calls == e.failOn {
		return nil, errEmbedTimeout
	}
	return fakeEmbedder{dim: e.dim}.Embed(text)
}

// zeroVectorEmbedder is the dishonest provider: it "succeeds" but hands back a
// vector carrying no signal. The write-path guard is the belt to R2's braces.
type zeroVectorEmbedder struct{ dim int }

func (z zeroVectorEmbedder) Embed(string) ([]float64, error) {
	return make([]float64, z.dim), nil
}

// TestIngestRoundTrip drives the real Rust extractor + Lance store end to end
// with a fake embedder over a temp-directory store.
func TestIngestRoundTrip(t *testing.T) {
	store, err := core.Open(t.TempDir(), "")
	if err != nil {
		t.Fatalf("open store: %v", err)
	}
	defer store.Close()

	emb := fakeEmbedder{dim: 8}
	data := []byte("Hello CiteNexus.\n\nSecond paragraph carries different words entirely.")

	rows, err := Ingest(store, data, "plain", "docING", emb)
	if err != nil {
		t.Fatalf("ingest: %v", err)
	}
	if len(rows) == 0 {
		t.Fatal("ingest produced no rows")
	}
	for _, r := range rows {
		if r.EuID == "" || len(r.Vector) != emb.dim || r.Text == "" {
			t.Fatalf("malformed row: %+v", r)
		}
	}

	// The rows really landed in the Rust store: scan them back.
	scanOut := store.Scan(-1)
	if strings.Contains(scanOut, `"error"`) {
		t.Fatalf("scan error: %s", scanOut)
	}
	var scanned []map[string]any
	if err := json.Unmarshal([]byte(scanOut), &scanned); err != nil {
		t.Fatalf("scan output not JSON: %v\n%s", err, scanOut)
	}
	if len(scanned) != len(rows) {
		t.Fatalf("scan returned %d rows, ingested %d", len(scanned), len(rows))
	}

	// A search against the first row's vector finds it back through the engine.
	vecJSON, err := json.Marshal(rows[0].Vector)
	if err != nil {
		t.Fatalf("marshal vector: %v", err)
	}
	searchOut := store.Search(string(vecJSON), 5)
	if strings.Contains(searchOut, `"error"`) {
		t.Fatalf("search error: %s", searchOut)
	}
	var found []map[string]any
	if err := json.Unmarshal([]byte(searchOut), &found); err != nil {
		t.Fatalf("search output not JSON: %v\n%s", err, searchOut)
	}
	if len(found) == 0 {
		t.Fatalf("search found nothing: %s", searchOut)
	}
}

// TestIngestRefusesWhenEmbedFails is the regression for ADR-0014 R2. The spike
// (spikes/model-seam-contract/go) proved that with the old no-error seam a model
// timing out on chunk 2 gave `err = <nil>`, the same row count as a healthy run,
// and a poisoned EU scoring cosine 0.0000 with no flag. Now ingest refuses and
// the store stays empty: fail-closed, all-or-nothing.
func TestIngestRefusesWhenEmbedFails(t *testing.T) {
	store, err := core.Open(t.TempDir(), "")
	if err != nil {
		t.Fatalf("open store: %v", err)
	}
	defer store.Close()

	data := []byte("The employee may disclose the defect.\n\nThe employee may NOT disclose the defect.")
	emb := &timeoutEmbedder{dim: 8, failOn: 2}

	rows, err := Ingest(store, data, "plain", "docFAIL", emb)
	if err == nil {
		t.Fatal("ingest returned nil error after the model failed — the corpus is silently wrong")
	}
	if !errors.Is(err, errEmbedTimeout) {
		t.Fatalf("the model's own error must reach the caller; got %v", err)
	}
	if rows != nil {
		t.Fatalf("ingest returned %d rows on failure; it must return none", len(rows))
	}
	assertStoreEmpty(t, store)
}

// TestIngestRefusesTheZeroVector: even an embedder that reports success must not
// be able to write a signal-free vector into the corpus.
func TestIngestRefusesTheZeroVector(t *testing.T) {
	store, err := core.Open(t.TempDir(), "")
	if err != nil {
		t.Fatalf("open store: %v", err)
	}
	defer store.Close()

	rows, err := Ingest(store, []byte("Hello CiteNexus."), "plain", "docZERO", zeroVectorEmbedder{dim: 8})
	if err == nil {
		t.Fatal("the zero vector was indexed")
	}
	if !strings.Contains(err.Error(), "zero vector") {
		t.Fatalf("error should name the zero vector: %v", err)
	}
	if rows != nil {
		t.Fatalf("ingest returned %d rows on failure; it must return none", len(rows))
	}
	assertStoreEmpty(t, store)
}

// assertStoreEmpty fails unless nothing at all landed in the Lance store.
func assertStoreEmpty(t *testing.T, store *core.Store) {
	t.Helper()
	scanOut := store.Scan(-1)
	if strings.Contains(scanOut, `"error"`) {
		// A store that was never written to may not have a table yet; that is
		// also "nothing landed".
		return
	}
	var scanned []map[string]any
	if err := json.Unmarshal([]byte(scanOut), &scanned); err != nil {
		t.Fatalf("scan output not JSON: %v\n%s", err, scanOut)
	}
	if len(scanned) != 0 {
		t.Fatalf("failed ingest wrote %d rows into the store: %s", len(scanned), scanOut)
	}
}

File: js/src/ingest/ingest.test.ts
// Ingest orchestrator test — exercises the full extract -> chunk -> embed ->
// store path, with the real Rust core (extract + Lance store) and a
// deterministic fake embedder. Requires the cdylib:
//   cd rust && cargo build --release

import { describe, it, expect, afterEach } from "vitest";
import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

import { Store } from "../core/core.js";
import { FakeEmbedding, EMBED_DIM } from "../fakes/fakes.js";
import { ingest } from "./ingest.js";

describe("ingest orchestrator", () => {
  const tmpDirs: string[] = [];
  afterEach(() => {
    for (const d of tmpDirs.splice(0)) rmSync(d, { recursive: true, force: true });
  });

  it("extracts, chunks, embeds and stores retrievable evidence units", async () => {
    const dir = mkdtempSync(join(tmpdir(), "citenexus-ingest-"));
    tmpDirs.push(dir);
    const store = Store.open(dir);
    const fake = new FakeEmbedding();
    const embed = (text: string): number[] => fake.embed(text);

    try {
      const text = "The cat sat on the mat.\n\nThe dog ran in the park.";
      const result = await ingest(store, new TextEncoder().encode(text), "plain", "docX", embed);

      expect(result.documentId).toBe("docX");
      expect(result.unitCount).toBeGreaterThan(0);
      expect(result.euIds).toHaveLength(result.unitCount);
      // eu_id shape: {document}::{order}::{chunk}
      expect(result.euIds[0]).toMatch(/^docX::\d+::\d+$/);

      // Everything ingested is scannable.
      const rows = store.scan();
      expect(rows).toHaveLength(result.unitCount);
      const dims = rows.map((r) => (r["vector"] as number[]).length);
      expect(new Set(dims)).toEqual(new Set([EMBED_DIM]));

      // The stored vectors are the ones we embedded, so the same query vector
      // retrieves its own unit at distance 0.
      const first = rows[0]!;
      const hits = store.search(first["vector"] as number[], 1);
      expect(hits).toHaveLength(1);
      expect(hits[0]!["eu_id"]).toBe(first["eu_id"]);

      // Re-ingesting the same document is idempotent (merge on eu_id).
      const again = await ingest(store, new TextEncoder().encode(text), "plain", "docX", embed);
      expect(again.unitCount).toBe(result.unitCount);
      expect(store.scan()).toHaveLength(result.unitCount);
    } finally {
      store.close();
    }
  });
  it("refuses — and writes NOTHING — when the model call fails", async () => {
    // The regression for ADR-0014 R2. With the old synchronous seam a failing
    // embedder had nowhere to say so; the rejection was never awaited and
    // `ingest()` returned a healthy-looking result over a poisoned corpus.
    const dir = mkdtempSync(join(tmpdir(), "citenexus-ingest-fail-"));
    tmpDirs.push(dir);
    const store = Store.open(dir);
    const fake = new FakeEmbedding();
    let calls = 0;
    const embed = async (text: string): Promise<number[]> => {
      calls += 1;
      if (calls === 2) throw new Error("model call timed out after 30s");
      return fake.embed(text);
    };

    try {
      const bytes = new TextEncoder().encode(
        "The employee may disclose the defect.\n\nThe employee may NOT disclose the defect.",
      );
      await expect(ingest(store, bytes, "plain", "docFAIL", embed)).rejects.toThrow(/timed out/);
      // Fail-closed and all-or-nothing: not one row landed.
      expect(store.scan()).toHaveLength(0);
    } finally {
      store.close();
    }
  });

  it("refuses the zero vector even when the embedder reports success", async () => {
    const dir = mkdtempSync(join(tmpdir(), "citenexus-ingest-zero-"));
    tmpDirs.push(dir);
    const store = Store.open(dir);
    const embed = (): number[] => new Array<number>(EMBED_DIM).fill(0);

    try {
      const bytes = new TextEncoder().encode("The cat sat on the mat.");
      await expect(ingest(store, bytes, "plain", "docZERO", embed)).rejects.toThrow(/zero vector/);
      expect(store.scan()).toHaveLength(0);
    } finally {
      store.close();
    }
  });
});
